doc/decisions/error_codes.md

```diff
@@ -0,0 +1,126 @@
+  [16 codes](https://godoc.org/google.golang.org/grpc/codes) defined. Some of these errors are similar or identical to those which will be used in elektra.
+  Every error of etcd is associated with one of these categories and gets its own error message which is specified in [this](https://github.com/etcd-io/etcd/blob/master/etcdserver/api/v3rpc/rpctypes/error.go) file. This concept though does not allow easy subclassing which might be useful (eg. further split FailedPrecondition into more specific errors like semantic and syntactic errors)
+- [Windows Registry](https://docs.microsoft.com/en-us/windows/desktop/sysinfo/registry):
+  The registry does not use any specific error concept but takes the standard [Win32 Error Codes](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/18d8fbe8-a967-4f1c-ae50-99ca8e491d2d). These are neither hierarchical nor have any special ordering. Basically it is the same as elektra has now except for no duplicated
+  errors.
+- macOS X plist:
+  Just like Windows, plist uses standard macOS X errors which is a [huge catalog](http://krypted.com/lists/comprehensive-list-of-mac-os-x-error-codes/) of unordered
+  return codes as integers.
+- [SNMP Standard](http://www.snmp.com/protocol/):
+  Being a standard network protocol, error codes are very specific to the domain itself. A list can be found [here](https://docs.microsoft.com/en-us/windows/desktop/snmp/snmp-error-codes) and would not meet the needs of elektra at all.
+- POSIX:
+  Returning a non-zero value and retrieving the concrete information from `errno` would not suffice for elektra as it is too simple. It would not solve any of our current
+  problems like having excessive uncategorized codes for errors.
+
+## Decision
+
+All "fatal" errors will be converted to "errors" as the distinguishment is not relevant.
+
+Unused marked errors will be removed from the specification.
+
+Errors will be categorizes into logical groups with subgroups.
+Each error will be made up of 5 characters, where the first 2 character indicate the highest level
+and character 3 to 5 will be used for subgrouping. Errors are prepended with the letter `C` which
+is the abbreviation for "Code".
+
+- Permanent errors
+  - Resource
+    - Memory Allocation C01110
+    - General Resource C01120
+  - Installation C01300
+  - Logical
+    - Assertion C01410
+    - Interface C01420
+    - Broken plugin C01430
+- Conflicting State C02000
+- Timeout C03000
+- Out-Of-Range C05000
+- Validation
+  - Syntactic C04100
+  - Semantic C04200
+
+## Rationale
+
+The grouping of errors will allow developers to filter for specific as well as more general errors to correctly
+react to them programmatically.
+The new concept will permit additional subgrouping of errors in case it might be needed in the future. Imagine the case where
+"Installation" errors is too general because developers saw a need for splitting the errors in "elektra installation"
```

Comment:
This is now about the semantics of "Installation errors" but this document should only be about error codes?